### ISSUE TITLE
pyg Implementation of GPSE

### ISSUE THREAD

**Issue by mossishahi** [G-Taxonomy-Workgroup/GPSE#4]
Hey everyone, 

I'm a bit confused with the [pyg implementation](https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.models.GPSE.html?highlight=gpse) of GPSE model. asking here to clarify for myself and maybe others.

Unfortunately, there is no full model training example of GPSE available under pyg so that I can trace the whole story and only a pretrained model is used there. However, your implementation is also limtied to those who love using graphgym, otherwise no choice!

AFAIK, `GPSE` passes the graphs (optionally with random node features) through a stack of `ResGatedGCN`s and then decodes different kind of `PSE`s on top of the latent space.

I guess the possible options are `ElstaticPE+LapPE+RWSE+HKdiagSE` for Node and `EigVals+CycleGE` for Graph. 

1. Then, I'm surprised why `num_node_targets` is set to 51 which has also ended to **51** `MLP`s each of which has only **1** output . does GPSE model needs one MLP per element? what I'm expecting is that I should have 4 MLPs given `ElstaticPE+LapPE+RWSE+HKdiagSE` each of which reconstructed by one MLP. 
2. There is only 1 `MLP` for graph level PSE which I was expecting to be 4 corresponding to all 4 PSEs. Surprisingly, the dimension of its output is **11** and I'm wondering how it is consistent to 51 for node level. 
4. I've created the GPSE model using pyg and it ended up in the following architecture. do you approve it's a right one?
 
```
GPSE(
  (pre_mp): GeneralMultiLayer(
    (Layer_0): GeneralLayer(
      (layer): Linear(
        (model): Linear(in_features=20, out_features=512, bias=False)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
  )
  (mp): GNNStackStage(
    (layer0): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer1): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer2): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer3): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer4): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer5): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer6): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer7): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer8): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer9): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer10): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer11): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer12): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer13): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer14): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer15): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer16): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer17): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer18): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
    (layer19): GeneralLayer(
      (layer): ResGatedGCNConv(
        (model): ResGatedGraphConv(512, 512)
      )
      (post_layer): Sequential(
        (0): BatchNorm1d(512, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
        (1): Dropout(p=0.2, inplace=False)
        (2): ReLU()
      )
    )
  )
  (post_mp): GNNInductiveHybridMultiHead(
    (node_post_mps): ModuleList(
      (0-50): 51 x MLP(
        (model): Sequential(
          (0): GeneralMultiLayer(
            (Layer_0): GeneralLayer(
              (layer): Linear(
                (model): Linear(in_features=512, out_features=32, bias=True)
              )
              (post_layer): Sequential(
                (0): Dropout(p=0.2, inplace=False)
                (1): ReLU()
              )
            )
          )
          (1): Linear(
            (model): Linear(in_features=32, out_features=1, bias=True)
          )
        )
      )
    )
    (graph_post_mp): MLP(
      (model): Sequential(
        (0): GeneralMultiLayer(
          (Layer_0): GeneralLayer(
            (layer): Linear(
              (model): Linear(in_features=512, out_features=512, bias=True)
            )
            (post_layer): Sequential(
              (0): Dropout(p=0.2, inplace=False)
              (1): ReLU()
            )
          )
        )
        (1): Linear(
          (model): Linear(in_features=512, out_features=11, bias=True)
        )
      )
    )
  )
)

```

I would really appreciate your response and would be happy that in case you don't approve this implementation, please link to a part of the your implementation which trains a GPSE model from scratch. I mean some code which shows `loss` function, `forward` path, etc.

**Comment by semihcanturk** [comment]
Hey @mossishahi , thanks for your interest in GPSE! So for starters, the current PyG implementation in GPSE is encoder-only, i.e. it is for pre-trained usage. I hope to get around to implementing the decoder in PyG this summer to enable end-to-end training here, but I have no real timeline on that -- so for the time being I suggest using **this repo** for pre-training your own GPSE models. From the README.md:

> You can also pre-train the GPSE model from scratch using the configs provided, e.g.
> ```python main.py --cfg configs/pretrain/gpse_molpcba.yaml```

Running this will output a model summary that you can compare with your output above. I would suggest using a debugger to step through the model building and training process in the codebase. To give you some pointers regarding your Q3:

- The model building process is basically handled by [model_builder.py](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/graphgym/model_builder.html) in GraphGym, which pulls up the relevant registered components (GNN layers, activations, pre/post message-passing MLPs, prediction heads) to build the GNN. Check out the `gnn` section of the [pretraining config](https://github.com/G-Taxonomy-Workgroup/GPSE/blob/main/configs/pretrain/gpse_molpcba.yaml) for details.
- In the config, you'll be interested in the [`inductive_hybrid_multi` head](https://github.com/G-Taxonomy-Workgroup/GPSE/blob/main/graphgym/head/inductive_hybrid.py) which is responsible for the PSE predictions. I think looking into how the head is constructed per PSE during runtime will help clarify things, though it seems you're already using it in your posted model output. 

So, I think your overall architectural build seems correct, though I strongly suggest comparing against the model output by running the code above as a sanity check, making sure all layers, number of MLP heads etc. match. Now, moving on to the specifics of the PSE predictions (Q1&2):

1. `num_node_targets` is equal to the total number of scalars predicted, not the number of PSE 'types', so to say. so what ends up actually happening is: 4 LapPE + 20 RWSE + 20 HKSE + 7 ElstaticPE = 51, where each MLP predicts only one of them. 1-MLP-per-scalar is one of several strategies we considered, e.g. a large MLP to predict all 51 scalars, or 1-MLP-per-PSE-type, but we just found that this works the best, and the MLPs are small enough that # parameters is still reasonable.
2. For graph-level PSEs, we have 4 Laplacian eigenvalues + cycle counts from 2-8, so 7+4=11. Again I believe we tried and found that for graph-level features a single MLP is sufficient, but again other approaches are also viable. Feel free to experiment with different heads, for our pre-training tasks these were optimal but if you're using datasets of different size & domains perhaps you may find more benefit from a different configuration (though frankly I'd expect the potential advantages to be somewhat marginal).

Do let me know if you have any questions, and given your work on this, do let me know if you want to implement the decoder in PyG and make a PR there -- I'd be happy to get more involved in helping out or reviewing.
